### DISCUSSION TITLE
Delay finishes early when another command is issued

### DISCUSSION THREAD

**Discussion by jan-kiszka** [arendst/Tasmota#20963] [Q&A] [answered]
Hi,

I'm trying to program a delayed power off via `backlog delay 3000; power1 off`. That works, provided no other command is issued during the delay phase (in my case, there are `status` commands issued from time to time). If so, delay finishes early, and the power-off is immediately executed. Is this a limitation or even an intended behavior of tasmota? I'm on 13.4.0.

Thanks,
Jan

**Comment by barbudor** [comment, marked as answer]
Forget about delay
Look at using RuleTimer

**Reply by sfromis** [reply]
For your specific use case of delayed power off, you could also use commands `TimedPower 300000,on` (will power off after millisconds) or `PulseTime` (to preset a duration used for every power on after that).

**Reply by jan-kiszka** [reply]
RuleTimer is indeed best for me as I can also cancel it easily. Just some feedback on invalid syntax would have been nice (`endon`, not `enddo`...).

**Reply by Jason2866** [reply]
There is no room for extensive syntax checks. Keep in mind everything is limited.

**Reply by jan-kiszka** [reply]
There is already some form of check that prevented my mistyped command to be executed. This could have simply printed a short notice, and the it would have been much clearer. That room is surely there.

**Reply by sfromis** [reply]
If Tasmota was to provide helpful error messages when not doing things the right way, that would take a *lot* of room, as there are lots of places where the result of bad input is that it will not do anything, or do something wrong.

**Reply by jan-kiszka** [reply]
This would have likely already helped and costed only a few dozens of bytes extra:
```
diff --git a/tasmota/tasmota_xdrv_driver/xdrv_10_rules.ino b/tasmota/tasmota_xdrv_driver/xdrv_10_rules.ino
index 903262d5a..3c6ccc523 100644
--- a/tasmota/tasmota_xdrv_driver/xdrv_10_rules.ino
+++ b/tasmota/tasmota_xdrv_driver/xdrv_10_rules.ino
@@ -733,16 +733,16 @@ bool RuleSetProcess(uint8_t rule_set, String &event_saved)
 
     String rule = rules;
     rule.toUpperCase();                                   // "ON INA219#CURRENT>0.100 DO BACKLOG DIMMER 10;COLOR 100000 ENDON"
-    if (!rule.startsWith(F("ON "))) { return serviced; }     // Bad syntax - Nothing to start on
+    if (!rule.startsWith(F("ON "))) { goto bad_syntax; }     // Bad syntax - Nothing to start on
 
     int pevt = rule.indexOf(F(" DO "));
-    if (pevt == -1) { return serviced; }                  // Bad syntax - Nothing to do
+    if (pevt == -1) { goto bad_syntax; }                  // Bad syntax - Nothing to do
     String event_trigger = rule.substring(3, pevt);       // "INA219#CURRENT>0.100"
     event_trigger.trim();
 
     plen = rule.indexOf(F(" ENDON"));
     plen2 = rule.indexOf(F(" BREAK"));
-    if ((plen == -1) && (plen2 == -1)) { return serviced; } // Bad syntax - No ENDON neither BREAK
+    if ((plen == -1) && (plen2 == -1)) { goto bad_syntax; } // Bad syntax - No ENDON neither BREAK
 
     if (plen == -1) { plen = 9999; }
     if (plen2 == -1) { plen2 = 9999; }
@@ -838,6 +838,10 @@ bool RuleSetProcess(uint8_t rule_set, String &event_saved)
     Rules.trigger_count[rule_set]++;
   }
   return serviced;
+
+bad_syntax:
+  AddLog(LOG_LEVEL_ERROR, PSTR("RUL: bad syntax: \"%s\""), commands.c_str());
+  return serviced;
 }
 
 /*******************************************************************************************/
```
(disclaimer: I'm not familiar with the code base, and this is not even compile-tested)

**Reply by Jason2866** [reply]
Too much ;-)

**Reply by jan-kiszka** [reply]
Unlikely, given how chatty this command is when it successfully runs.

**Reply by Jason2866** [reply]
If you manage to do a PR with zero increase in code size (saving bytes in the success case) it may have a chance to get merged when not having breaking changes.

**Reply by sfromis** [reply]
The "problem" is not so much adding one single check, but that **lots** would be needed to be consistent about detecting errors gracefully.

**Comment by Wintespe** [comment]
I could well imagine writing a small external tool for Windows that would check the syntax of a rule. But I would need an EBNF for the rules or something like that. With the current documentation this is quite time consuming and error prone.
